x-pack/plugins/lens/public/datasources/text_based/components/dimension_editor.tsx

```diff
@@ -0,0 +1,111 @@
+import { retrieveLayerColumnsFromCache, getColumnsFromCache } from '../fieldlist_cache';
+
+export type TextBasedDimensionEditorProps =
+  DatasourceDimensionEditorProps<TextBasedPrivateState> & {
+    expressions: ExpressionsStart;
+  };
+
+export function TextBasedDimensionEditor(props: TextBasedDimensionEditorProps) {
+  const query = props.state.layers[props.layerId]?.query;
+
+  const allColumns = retrieveLayerColumnsFromCache(
+    props.state.layers[props.layerId]?.columns ?? [],
+    query
+  );
+  const allFields = query ? getColumnsFromCache(query) : [];
+  const hasNumberTypeColumns = allColumns?.some((c) => c?.meta?.type === 'number');
+  const fields = allFields.map((col) => {
+    return {
+      id: col.id,
+      name: col.name,
+      meta: col?.meta ?? { type: 'number' },
+      compatible:
+        props.isMetricDimension && hasNumberTypeColumns
+          ? props.filterOperations({
+              dataType: col?.meta?.type as DataType,
+              isBucketed: Boolean(col?.meta?.type !== 'number'),
+              scale: 'ordinal',
+            })
+          : true,
+    };
+  });
+  const selectedField = allColumns?.find((column) => column.columnId === props.columnId);
+
+  return (
+    <>
+      <EuiFormRow
+        data-test-subj="text-based-languages-field-selection-row"
+        label={i18n.translate('xpack.lens.textBasedLanguages.chooseField', {
+          defaultMessage: 'Field',
+        })}
+        fullWidth
+        className="lnsIndexPatternDimensionEditor--padded"
+      >
+        <FieldSelect
+          existingFields={fields ?? []}
+          selectedField={selectedField}
+          onChoose={(choice) => {
+            const meta = fields?.find((f) => f.name === choice.field)?.meta;
+            const newColumn = {
+              columnId: props.columnId,
+              fieldName: choice.field,
+              meta,
+            };
+            return props.setState(
+              !selectedField
+                ? {
+                    ...props.state,
+                    layers: {
+                      ...props.state.layers,
+                      [props.layerId]: {
+                        ...props.state.layers[props.layerId],
+                        columns: [...props.state.layers[props.layerId].columns, newColumn],
+                      },
+                    },
+                  }
+                : {
+                    ...props.state,
+                    layers: {
+                      ...props.state.layers,
+                      [props.layerId]: {
+                        ...props.state.layers[props.layerId],
+                        columns: props.state.layers[props.layerId].columns.map((col) =>
+                          col.columnId !== props.columnId
+                            ? col
+                            : { ...col, fieldName: choice.field, meta }
+                        ),
+                      },
+                    },
+                  }
+            );
```

Comment:
Can we push down the condition at `columns` level?

```suggestion
                {
                    ...props.state,
                    layers: {
                      ...props.state.layers,
                      [props.layerId]: {
                        ...props.state.layers[props.layerId],
                        columns: 
              !selectedField ? [...props.state.layers[props.layerId].columns, newColumn] : props.state.layers[props.layerId].columns.map((col) =>
                          col.columnId !== props.columnId
                            ? col
                            : { ...col, fieldName: choice.field, meta }
                        ),
                      },
                    }